---
title: "Data analysis tutorial for BT8121"
knit: (function(input_file, encoding) {
  out_dir <- 'docs';
  rmarkdown::render(input_file,
 encoding=encoding,
 output_file=file.path(dirname(input_file), out_dir, 'index.html'))})
author: "[Eirini Tsirvouli]()"
date: "Last updated: `r format(Sys.time(), '%d %B, %Y')`"
output:
  html_document:
    css: style.css
    theme: united
    toc: true
    toc_float:
      collapsed: false
      smooth_scroll: true
    toc_depth: 3
    number_sections: false
    code_folding: hide
    code_download: true
bibliography: references.bib
link-citations: true
---

# Intro {-}
In this document, a tutorial for a multi-omics data analysis of cancer data is provided. The tutorial was composed for the students of BT8121 in Norwegian University of Science and Technology and it aims to their training in data retrieval and analysis as a starting point of logical model building. All data used for this tutorial are public and retrieved from [GDC Data Portal](https://portal.gdc.cancer.gov/). The tutorial makes use of some existing packages and parts of their vignettes, such as [TCGABioLinks](https://bioconductor.org/packages/release/bioc/html/TCGAbiolinks.html). 

The markdown document was created using the template from [rtemps](https://github.com/bblodfon/rtemps) package.


# Load libraries {-}

Loading libraries:
```{r Load libraries, message = FALSE, echo=FALSE}
library(DT)
library(ggplot2)
library(xfun)
library(TCGAbiolinks)
library(tidyverse)
library(maftools)
library(data.table)
library(knitr)
opts_chunk$set(tidy.opts=list(width.cutoff=60),tidy=TRUE)
```

# Query expression data {-}

You can query GDC data using the `GDCquery` function. Each cancer project of TCGA can be identified by a unique name that should be specified in the function. For Breast Cancer the ID is TCGA-BRCA and for Acute Myeloid Leukemia it is TCGA-LAML. For more information on the used arguments please run `?GDCquery`.

To explore the data categories available for each project run `TCGAbiolinks:::getProjectSummary("project")`

For the rest of the tutorial the TCGA-BRCA will be selected.

```{r, eval = FALSE}
query <- GDCquery(project = "TCGA-BRCA", data.category = "Transcriptome Profiling", data.type = "Gene Expression Quantification", workflow.type = "HTSeq - Counts")

GDCdownload(query)

data <- GDCprepare(query)
```

## Preprocessing {-}

Prepare expression matrix with geneID in the rows and samples (barcode) in the columns

```{r, eval = FALSE}
BRCARnaseq_assay <- GDCprepare(query)

BRCAMatrix <- assay(BRCARnaseq_assay,"raw_counts")
```

For gene expression if you need to see a boxplot correlation and AAIC plot to define outliers you can run

```{r, eval = FALSE}
BRCARnaseq_CorOutliers <- TCGAanalyze_Preprocessing(BRCARnaseq_assay)`
```

### Differential expression analysis (DEA) {-}

Using the `TCGAanalyze_DEA` function, a Differential Expression Analysis can be performed in order to idenify differentially expressed genes (DEGs) between normal and tumor samples. The `TCGAanalyze_DEA` uses fucntions of the [edgeR package](https://bioconductor.org/packages/release/bioc/html/edgeR.html).


Perform DEA (Differential expression analysis) to identify differentially expressed genes (DEGs) using the TCGAanalyze_DEA function.

```{r, eval = FALSE}

# normalization of genes
dataNorm <- TCGAanalyze_Normalization(tabDF = dataBRCA, geneInfo =  geneInfo)

# quantile filter of genes
dataFilt <- TCGAanalyze_Filtering(tabDF = dataNorm,
                                  method = "quantile", 
                                  qnt.cut =  0.25)

# selection of normal samples "NT"
samplesNT <- TCGAquery_SampleTypes(barcode = colnames(dataFilt),
                                   typesample = c("NT"))

# selection of tumor samples "TP"
samplesTP <- TCGAquery_SampleTypes(barcode = colnames(dataFilt), 
                                   typesample = c("TP"))

# Diff.expr.analysis (DEA)
dataDEGs <- TCGAanalyze_DEA(mat1 = dataFilt[,samplesNT],
                            mat2 = dataFilt[,samplesTP],
                            Cond1type = "Normal",
                            Cond2type = "Tumor",
                            fdr.cut = 0.01 ,
                            logFC.cut = 1,
                            method = "glmLRT")

# DEGs table with expression values in normal and tumor samples
dataDEGsFiltLevel <- TCGAanalyze_LevelTab(dataDEGs,"Tumor","Normal",
                                          dataFilt[,samplesTP],dataFilt[,samplesNT])

```

# Query mutation data {-}

```{r, eval=FALSE}
mut <- GDCquery_Maf(tumor = "BRCA", pipelines = "mutect2", save.csv = TRUE)

TCGAvisualize_oncoprint(mut = mut, genes = mut$Hugo_Symbol[1:20],
                        filename = "oncoprint.pdf",
                        color=c("background"="#CCCCCC","DEL"="purple","INS"="yellow","SNP"="brown"),
                        rows.font.size=10,
                        width = 5,
                        heatmap.legend.side = "right",
                        dist.col = 0,
                        label.font.size = 10)
```

# Query Copy Number Variation data {-}

```{r, eval=FALSE}
query <- GDCquery(project = "TCGA-BRCA",
                        data.category = "Copy Number Variation",
                        data.type = "Masked Copy Number Segment",
                        sample.type = c("Primary solid Tumor"))

GDCdownload(query)

GDCprepare(query, save = TRUE, save.filename = "BRCA_maskedCNV.rda")
load("BRCA_maskedCNV.rda")

```

## GISTIC 2.0 analysis {-}

The [GISTIC 2.0](https://www.genepattern.org/modules/docs/GISTIC_2.0) (Genomic Identification of Significant Targets in Cancer) tool can be used to identify genomic regions that are significantly amplified or deleted in the cancer type of interest. GISTIC 2.0 can be run through the GenePattern server or locally from your computer. In this tutorial GenePattern is used.

## Prepare CNV data for GISTIC analysis

GISTIC requires specific format as an input (for more information please refer to the [Documentation](http://portals.broadinstitute.org/cgi-bin/cancer/publications/pub_paper.cgi?mode=view&paper_id=216&p=t)). The following code can be used to prepare your data in the correct format.

```{r, eval=FALSE}
cnvMatrix = data[,-1]
colnames(cnvMatrix) <- c("Chromosome", "Start Position", "End Position", "Num markers", "Seg.CN","Sample")
setcolorder(cnvMatrix, c("Sample","Chromosome","Start Position", "End Position", "Num markers", "Seg.CN"))
str(cnvMatrix)
cnvMatrix[cnvMatrix$Chromosome == "X","Chromosome"] <- 23
cnvMatrix[cnvMatrix$Chromosome == "Y","Chromosome"] = 24
cnvMatrix$Chromosome <- as.integer(cnvMatrix$Chromosome)
unique(cnvMatrix$Chromosome)

cnvMatrix$`Start Position`= as.integer(cnvMatrix$`Start Position`)
cnvMatrix$`End Position` = as.integer(cnvMatrix$`End Position`)
cnvMatrix$`Num markers` = as.integer(cnvMatrix$`Num markers`)

cnvMatrix$test = cnvMatrix$`Start Position` - cnvMatrix$`End Position`
GISTIC_BRCA = subset(cnvMatrix, cnvMatrix$test < 0)
GISTIC_BRCA = GISTIC_BRCA[,-7]

write.table(GISTIC_BRCA, file = "GISTIC_BRCA.txt", quote = FALSE, row.names = FALSE, sep = "\t" )
```

### Identify genes from GISTIC results

The analysis returns the recurrently amplified or deleted peaks and their associated q-value. The following code can be used to convert the results into a list of recurrently amplified and deleted genes.

```{r, eval=FALSE}
BRCA_gistic = readGistic(gisticAllLesionsFile = "BRCA_gistic.all_lesions.conf_99.txt",
                         gisticAmpGenesFile = "BRCA_gistic.amp_genes.conf_99.txt",
                         gisticDelGenesFile = "BRCA_gistic.del_genes.conf_99.txt",
                         gisticScoresFile = "BRCA_gistic.scores.gistic")


#Visualize your results
gisticChromPlot(gistic = BRCA_gistic)
gisticBubblePlot(gistic = BRCA_gistic)
gisticOncoPlot(gistic = BRCA_gistic)

# Translate gene names
BRCA_gistic_genes_HUGO = as.data.frame(BRCA_gistic@gene.summary$Hugo_Symbol)
mart <- useDataset("hsapiens_gene_ensembl", useMart("ensembl"))

BRCA_gistic.genes <- getBM(
  filters="hgnc_symbol",
  attributes=c("hgnc_symbol", "ensembl_gene_id", "entrezgene"),
  values=BRCA_gistic_genes_HUGO,
  mart=mart)

BRCA_gistic.genes = BRCA_gistic.genes[!duplicated(BRCA_gistic.genes$hgnc_symbol),]
```

# Query methylation data {-}

```{r, eval=FALSE}
query <- GDCquery(project = "TCGA-BRCA", 
                      data.category = "DNA Methylation")

GDCdownload(query)

BRCAmet <- GDCprepare(query = query,
                      save = TRUE, 
                      save.filename = "BRCAmet.rda",
                      summarizedExperiment = TRUE)
```

## Differentially methylated regions analysis {-}

```{r, eval=FALSE}


data <- TCGAanalyze_DMR(data, groupCol = "methylation_subtype",
                        group1 = "CIMP.H",
                        group2="CIMP.L",
                        p.cut = 10^-5,
                        diffmean.cut = 0.25,
                        legend = "State",
                        plot.filename = "coad_CIMPHvsCIMPL_metvolcano.png")
```


# Enrichment analysis {-}

In order to identify which are those processes the affected genes identified from your analysis, you can perform an Enrichment Analysis. The enrichment analysis can be performed against various databases and/or against GO terms.

Depending on your choice you can use one (or more) of the following packages for your analysis.

Reactome EA: [ReactomePA](http://bioconductor.org/packages/release/bioc/html/ReactomePA.html)
KEGG and GOterms:
ACSN:

# R session info {-}

```{r session info, comment=""}
xfun::session_info()
```

